Objects references and copying (#223)
javascript-tutorial/pt.javascript.info: 1-js/04-object-basics/02-object-copy/article.md

alert( user.name ); // ainda John no objeto original
```

Também podemos usar o método [Object.assign](mdn:js/Object/assign) para isso.

A sintaxe é:

```js run
Object.assign(dest, [fonte1, fonte2, fonte3...])
```

- O primeiro argumento `dest` é o objeto destino.
- Argumentos adicionais `fonte1, ..., fonteN` (pode ser quantos precisar) são os objetos fonte.
- Ele copia as propriedades de todos os objetos fontes `fonte1, ..., fonteN` para o destino `dest`. Em outras palavras, propriedaes de todos os argumentos começando pelo segundo são copiadas para o primeiro objeto.
- A chamada retorna `dest`.

Por exemplo, podemos usá-lo para fundir diversos objetos em um:
```js run
let user = { name: "John" };

let permissions1 = { canView: true };
let permissions2 = { canEdit: true };

*!*
// copia todas as propriedades de permissions1 e permissions2 para user 
Object.assign(user, permissions1, permissions2);
*/!*

// agora user = { name: "John", canView: true, canEdit: true }
```

Se o nome da propriedade copiada já existir, ela é sobrescrita:

```js run
let user = { name: "John" };

Object.assign(user, { name: "Pete" });

alert(user.name); // agora user = { name: "Pete" }
```

> Podemos também utilizar `Object.assign` para substituir `for..in` loop para clonagem simples:

Reviewer: ```suggestion
Podemos também utilizar `Object.assign` para substituir o ciclo `for..in`  para clonagem simples:
```

Reviewer: Please, you may also leave 'loop' without translation.

};

*!*
let clone = {}; // o novo objeto vazio

// vamos copiar todas as propriedades de user para ele
for (let key in user) {
  clone[key] = user[key];
}
*/!* 

// agora clone é um objeto totalmente independente com o mesmo conteúdo
clone.name = "Pete"; // alterada a informação nele

alert( user.name ); // ainda John no objeto original
```

> Também podemos usar o método [Object.assign](mdn:js/Object/assign) para isso.

Reviewer: ```suggestion
Também podemos usar o método [Object.assign](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) para isso.
```

A sintaxe é:

```js run
Object.assign(dest, [fonte1, fonte2, fonte3...])
```

- O primeiro argumento `dest` é o objeto destino.

> - Argumentos adicionais `fonte1, ..., fonteN` (pode ser quantos precisar) são os objetos fonte.

Reviewer: ```suggestion
- Argumentos adicionais `fonte1, ..., fonteN` (podem ser quantos precisar) são os objetos fonte.
```

> - Ele copia as propriedades de todos os objetos fontes `fonte1, ..., fonteN` para o destino `dest`. Em outras palavras, propriedaes de todos os argumentos começando pelo segundo são copiadas para o primeiro objeto.

Reviewer: ```suggestion
- Ele copia as propriedades de todos os objetos fontes `fonte1, ..., fonteN` para o destino `dest`. Por outras palavras, as propriedaes de todos os argumentos começando pelo segundo são copiadas para o primeiro objeto.
```

*/!*
```

Ele copia todas as propriedade de `user` para o objeto vazio e o retorna.

> Também há outros métodos para clonagem de objeto, por exemplo usando a [sintaxe espalhada](info:rest-parameters-spread) `clone = {...user}`, coberto mais tarde no tutorial.

Reviewer: ```suggestion
Também há outros métodos para clonagem de objeto, por exemplo usando a [sintaxe espalhada](info:rest-parameters-spread) `clone = {...user}`, coberta mais tarde no tutorial.
```

## Clonagem aninhada

> Ate agora assumimos que todas as propriedades de `user` são primitivas. Mas propriedades podem ser referências para outros objetos. O que fazer com essas propriedades?

Reviewer: ```suggestion
Até agora assumimos que todas as propriedades de `user` são primitivas. Mas propriedades podem ser referências para outros objetos. O que fazer com essas propriedades?
```

let clone = Object.assign({}, user);

> alert( user.sizes === clone.sizes ); // verdade, mesmo objeto

Reviewer: ```suggestion
alert( user.sizes === clone.sizes ); // true (verdade), mesmo objeto
```

// user e clone compartilham tamanhos

> user.sizes.width++;       // altera a propriedade por um

Reviewer: ```suggestion
user.sizes.width++;       // altere a propriedade em um lugar
```

> alert(clone.sizes.width); // 51, olhe o outro resultado

Reviewer: ```suggestion
alert(clone.sizes.width); // 51, veja o resultado no outro 
```

```

> Para concertar isso, precisamos usar um ciclo de clonagem para examinar cada valor de `user[key]` e, se for um objeto, então replicar também sua estrutura. Isso é chamado de "clonagem profunda".

Reviewer: ```suggestion
Para concertar isso, precisamos usar um ciclo de clonagem para examinar cada valor de `user[key]` e, se for um objeto, então replicar também a sua estrutura. Isso é chamado de "clonagem profunda".
```

Podemos implementar usando recursão. Ou, para não reinventar a roda, pegar uma implementação existente, por exemplo [_.cloneDeep(obj)](https://lodash.com/docs#cloneDeep) da biblioteca do JavaScript [lodash](https://lodash.com).

> ````smart header="Const objects can be modified"

Reviewer: ```suggestion
````smart header="Objetos constantes podem ser modificados"
```

> Um efeito colateral importante de armazenar objetos como referência é que objetos declarados como `const` podem ser modificados.

Reviewer: ```suggestion
Um efeito colateral importante de armazenar objetos como referência é que objetos declarados como `const` *podem* ser modificados.
```

alert(user.name); // Pete
```

> Pode parecer que a linha (*) causaria um erro, mas não causa. O valor de `user` é constante, precisa sempre referenciar o mesmo objeto, mas propriedades desse objeto são livres para serem alteradas.

Reviewer: ```suggestion
Pode parecer que a linha (*) causaria um erro, mas não causa. O valor de `user` é constante, precisa de sempre referenciar o mesmo objeto, mas propriedades desse objeto são livres de serem alteradas.
```

Em outras palavras, o `const user` dá um erro apenas se tentarmos definir `user=...` como um todo

> Dito isso, se realmente precisarmos criar propriedades constantes no objeto, também é possível, mas usando métodos totalmente diferentes. Iremos menconar isso no capítulo <info:property-descriptors>.

Reviewer: ```suggestion
Dito isso, se realmente precisarmos de criar propriedades constantes no objeto, também é possível, mas usando métodos totalmente diferentes. Iremos menconar isso no capítulo <info:property-descriptors>.
```

````

## Sumário

> Objetos são atribuidos e copiados por referência. Em outras palavras, uma variável armazena não o "valor do objeto", mas sim sua "referência" (endereço na memória) para o valor. Então copiar tal variável ou passar ela como um argumento de uma função copia a referência, não o proprio objeto.

Reviewer: ```suggestion
Objetos são atribuidos e copiados por referência. Por outras palavras, uma variável armazena não o "valor do objeto", mas sim uma "referência" (endereço na memória) para o valor. Então copiar tal variável ou passar ela como um argumento de uma função copia a referência, não o proprio objeto.
```

> Todas as operações por meio de referências copiadas(como adicionar/remover propriedades) são realizadas no mesmo objeto único.

Reviewer: ```suggestion
Todas as operações por meio de referências copiadas(como adicionar/remover propriedades) são realizadas no mesmo único objeto.
```

> Para criar uma "cópia real" (um clone) podemos usar `Object.assign` para o então chamado "cópia superficial" (objetos aninhados são copiados por referência) ou uma função de "clonagem profunda", como a [_.cloneDeep(obj)](https://lodash.com/docs#cloneDeep).

Reviewer: ```suggestion
Para criar uma "cópia real" (um clone) podemos usar `Object.assign` para a chamada "cópia superficial" (objetos aninhados são copiados por referência) ou uma função de "clonagem profunda", como a [_.cloneDeep(obj)](https://lodash.com/docs#cloneDeep).
```
